### ISSUE TITLE
Warnings from -Wformat in regcomp.c

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12939]
Migrated from [rt.perl.org#117837](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=117837) (status was 'resolved')

Searchable as RT117837$

**Comment by p5pRT** [comment]
### From rmbarker.cpan@btinternet.com

#### Created by rmbarker.cpan@btinternet.com



When compiling perl\, I get many \-Wformat warnings from regcomp\.c
%X format is used but the variable is not an int \(i\.e\. is longer\)
The normal fix for this is to use UVXf and cast the variable to \(UV\)
and this is what the attached patch does\.

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=low

Site configuration information for perl 5.17.12:

Configured by robin at Wed May  1 12:59:24 BST 2013.

Summary of my perl5 (revision 5 version 17 subversion 12) configuration:
   Commit id: 8257347d5f11b9bd1df15345570993538c8b1696
   Platform:
     osname=linux, osvers=3.0.0-32-generic, archname=i686-linux-64int
     uname='linux king-ubuntu 3.0.0-32-generic #51-ubuntu smp thu mar 21 
15:51:26 utc 2013 i686 athlon i386 gnulinux '
     config_args='-ders -Dcc=gcc -Dusedevel -Doptimize=-O2 -DDEBUGGING 
-Duse64bitint -Dman3dir=none -Dcf_email=rmbarker.cpan@btinternet.com'
     hint=recommended, useposix=true, d_sigaction=define
     useithreads=undef, usemultiplicity=undef
     useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
     use64bitint=define, use64bitall=undef, uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='gcc', ccflags ='-DDEBUGGING -fno-strict-aliasing -pipe 
-fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE 
-D_FILE_OFFSET_BITS=64',
     optimize='-O2 -g',
     cppflags='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector 
-I/usr/local/include'
     ccversion='', gccversion='4.6.4', gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
     ivtype='long long', ivsize=8, nvtype='double', nvsize=8, 
Off_t='off_t', lseeksize=8
     alignbytes=4, prototype=define
   Linker and Libraries:
     ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
     libpth=/usr/local/lib /lib/i386-linux-gnu /lib/../lib 
/usr/lib/i386-linux-gnu /usr/lib/../lib /lib /usr/lib
     libs=-lnsl -ldl -lm -lcrypt -lutil -lc
     perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
     libc=, so=so, useshrplib=false, libperl=libperl.a
     gnulibc_version='2.13'
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
     cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib 
-fstack-protector'

Locally applied patches:



@INC for perl 5.17.12:
     lib
     /usr/local/lib/perl5/site_perl/5.17.12/i686-linux-64int
     /usr/local/lib/perl5/site_perl/5.17.12
     /usr/local/lib/perl5/5.17.12/i686-linux-64int
     /usr/local/lib/perl5/5.17.12
     .


Environment for perl 5.17.12:
     HOME=/home/robin
     LANG=en_GB.UTF-8
     LANGUAGE=en_GB:en
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
PATH=/usr/lib/lightdm/lightdm:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
     PERL_BADLANG (unset)
     SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From rmbarker.cpan@btinternet.com

<details><summary>0002-deal-with-Wformat-warnings.patch</summary>

```diff
From 8257347d5f11b9bd1df15345570993538c8b1696 Mon Sep 17 00:00:00 2001
From: Robin Barker <rmbarker@cpan.org>
Date: Wed, 1 May 2013 18:45:07 +0100
Subject: [PATCH 2/2] deal with -Wformat warnings

---
 regcomp.c |   59 ++++++++++++++++++++++++++++++-----------------------------
 1 files changed, 30 insertions(+), 29 deletions(-)

diff --git a/regcomp.c b/regcomp.c
index 0840778..5146001 100644
--- a/regcomp.c
+++ b/regcomp.c
@@ -1135,7 +1135,7 @@ S_dump_trie(pTHX_ const struct _reg_trie_data *trie, HV *widecharmap,
         PerlIO_printf( Perl_debug_log, "%*s#%4"UVXf"|", (int)depth * 2 + 2,"", (UV)state);
 
         if ( trie->states[ state ].wordnum ) {
-            PerlIO_printf( Perl_debug_log, " W%4X", trie->states[ state ].wordnum );
+            PerlIO_printf( Perl_debug_log, " W%4" UVXf, (UV)trie->states[ state ].wordnum );
         } else {
             PerlIO_printf( Perl_debug_log, "%6s", "" );
         }
@@ -1216,14 +1216,14 @@ S_dump_trie_interim_list(pTHX_ const struct _reg_trie_data *trie,
         for( charid = 1 ; charid <= TRIE_LIST_USED( state ) ; charid++ ) {
 	    SV ** const tmp = av_fetch( revcharmap, TRIE_LIST_ITEM(state,charid).forid, 0);
 	    if ( tmp ) {
-                PerlIO_printf( Perl_debug_log, "%*s:%3X=%4"UVXf" | ",
+                PerlIO_printf( Perl_debug_log, "%*s:%3"UVXf"=%4"UVXf" | ",
                     colwidth,
                     pv_pretty(sv, SvPV_nolen_const(*tmp), SvCUR(*tmp), colwidth, 
 	                    PL_colors[0], PL_colors[1],
 	                    (SvUTF8(*tmp) ? PERL_PV_ESCAPE_UNI : 0) |
 	                    PERL_PV_ESCAPE_FIRSTCHAR 
                     ) ,
-                    TRIE_LIST_ITEM(state,charid).forid,
+                    (UV)TRIE_LIST_ITEM(state,charid).forid,
                     (UV)TRIE_LIST_ITEM(state,charid).newstate
                 );
                 if (!(charid % 10)) 
@@ -1292,15 +1292,16 @@ S_dump_trie_interim_table(pTHX_ const struct _reg_trie_data *trie,
         for( charid = 0 ; charid < trie->uniquecharcount ; charid++ ) {
             UV v=(UV)SAFE_TRIE_NODENUM( trie->trans[ state + charid ].next );
             if (v)
-                PerlIO_printf( Perl_debug_log, "%*"UVXf, colwidth, v );
+                PerlIO_printf( Perl_debug_log, "%*"UVXf, colwidth, (UV)v );
             else
                 PerlIO_printf( Perl_debug_log, "%*s", colwidth, "." );
         }
         if ( ! trie->states[ TRIE_NODENUM( state ) ].wordnum ) {
             PerlIO_printf( Perl_debug_log, " (%4"UVXf")\n", (UV)trie->trans[ state ].check );
         } else {
-            PerlIO_printf( Perl_debug_log, " (%4"UVXf") W%4X\n", (UV)trie->trans[ state ].check,
-            trie->states[ TRIE_NODENUM( state ) ].wordnum );
+            PerlIO_printf( Perl_debug_log, " (%4"UVXf") W%4"UVXf"\n",
+            (UV)trie->trans[ state ].check,
+            (UV)trie->states[ TRIE_NODENUM( state ) ].wordnum );
         }
     }
 }
@@ -5817,7 +5818,7 @@ Perl_re_op_compile(pTHX_ SV ** const patternp, int pat_count,
                                     pRExC_state->num_code_blocks);
             goto redo_first_pass;
         }
-        Perl_croak(aTHX_ "panic: reg returned NULL to re_op_compile for sizing pass, flags=%#X", flags);
+        Perl_croak(aTHX_ "panic: reg returned NULL to re_op_compile for sizing pass, flags=%#" UVXf, (UV)flags);
     }
     if (code_blocksv)
 	SvLEN_set(code_blocksv,0); /* no you can't have it, sv_clear */
@@ -5989,7 +5990,7 @@ Perl_re_op_compile(pTHX_ SV ** const patternp, int pat_count,
     REGC((U8)REG_MAGIC, (char*) RExC_emit++);
     if (reg(pRExC_state, 0, &flags,1) == NULL) {
 	ReREFCNT_dec(rx);   
-        Perl_croak(aTHX_ "panic: reg returned NULL to re_op_compile for generation pass, flags=%#X", flags);
+        Perl_croak(aTHX_ "panic: reg returned NULL to re_op_compile for generation pass, flags=%#" UVXf, (UV)flags);
     }
     /* XXXX To minimize changes to RE engine we always allocate
        3-units-long substrs field. */
@@ -9147,8 +9148,8 @@ S_reg(pTHX_ RExC_state_t *pRExC_state, I32 paren, I32 *flagp,U32 depth)
                             *flagp = RESTART_UTF8;
                             return NULL;
                         }
-                        FAIL2("panic: regbranch returned NULL, flags=%#X",
-                              flags);
+                        FAIL2("panic: regbranch returned NULL, flags=%#" UVXf,
+                              (UV)flags);
                     } else
                         REGTAIL(pRExC_state, br, reganode(pRExC_state, LONGJMP, 0));
 		    c = *nextchar(pRExC_state);
@@ -9163,8 +9164,8 @@ S_reg(pTHX_ RExC_state_t *pRExC_state, I32 paren, I32 *flagp,U32 depth)
                                 *flagp = RESTART_UTF8;
                                 return NULL;
                             }
-                            FAIL2("panic: regbranch returned NULL, flags=%#X",
-                                  flags);
+                            FAIL2("panic: regbranch returned NULL, flags=%#" UVXf,
+                                  (UV)flags);
                         }
                         REGTAIL(pRExC_state, ret, lastbr);
 		 	if (flags&HASWIDTH)
@@ -9252,7 +9253,7 @@ S_reg(pTHX_ RExC_state_t *pRExC_state, I32 paren, I32 *flagp,U32 depth)
             *flagp = RESTART_UTF8;
             return NULL;
         }
-        FAIL2("panic: regbranch returned NULL, flags=%#X", flags);
+        FAIL2("panic: regbranch returned NULL, flags=%#" UVXf, (UV)flags);
     }
     if (*RExC_parse == '|') {
 	if (!SIZE_ONLY && RExC_extralen) {
@@ -9297,7 +9298,7 @@ S_reg(pTHX_ RExC_state_t *pRExC_state, I32 paren, I32 *flagp,U32 depth)
                 *flagp = RESTART_UTF8;
                 return NULL;
             }
-            FAIL2("panic: regbranch returned NULL, flags=%#X", flags);
+            FAIL2("panic: regbranch returned NULL, flags=%#" UVXf, (UV)flags);
         }
         REGTAIL(pRExC_state, lastbr, br);               /* BRANCH -> BRANCH. */
 	lastbr = br;
@@ -9501,7 +9502,7 @@ S_regbranch(pTHX_ RExC_state_t *pRExC_state, I32 *flagp, I32 first, U32 depth)
                 *flagp = RESTART_UTF8;
                 return NULL;
             }
-            FAIL2("panic: regpiece returned NULL, flags=%#X", flags);
+            FAIL2("panic: regpiece returned NULL, flags=%#" UVXf, (UV)flags);
 	}
 	else if (ret == NULL)
 	    ret = latest;
@@ -9571,7 +9572,7 @@ S_regpiece(pTHX_ RExC_state_t *pRExC_state, I32 *flagp, U32 depth)
 	if (flags & (TRYAGAIN|RESTART_UTF8))
 	    *flagp |= flags & (TRYAGAIN|RESTART_UTF8);
         else
-            FAIL2("panic: regatom returned NULL, flags=%#X", flags);
+            FAIL2("panic: regatom returned NULL, flags=%#" UVXf, (UV)flags);
 	return(NULL);
     }
 
@@ -10013,8 +10014,8 @@ S_grok_bslash_N(pTHX_ RExC_state_t *pRExC_state, regnode** node_p, UV *valuep, I
                 *flagp = RESTART_UTF8;
                 return FALSE;
             }
-            FAIL2("panic: reg returned NULL to grok_bslash_N, flags=%#X",
-                  flags);
+            FAIL2("panic: reg returned NULL to grok_bslash_N, flags=%#" UVXf,
+                  (UV)flags);
         } 
 	*flagp |= flags&(HASWIDTH|SPSTART|SIMPLE|POSTPONED);
 
@@ -10291,8 +10292,8 @@ tryagain:
         if (ret == NULL) {
             if (*flagp & RESTART_UTF8)
                 return NULL;
-            FAIL2("panic: regclass returned NULL to regatom, flags=%#X",
-                  *flagp);
+            FAIL2("panic: regclass returned NULL to regatom, flags=%#" UVXf,
+                  (UV)*flagp);
         }
 	nextchar(pRExC_state);
         Set_Node_Length(ret, RExC_parse - oregcomp_parse + 1); /* MJD */
@@ -10314,7 +10315,7 @@ tryagain:
                     *flagp = RESTART_UTF8;
                     return NULL;
                 }
-                FAIL2("panic: reg returned NULL to regatom, flags=%#X", flags);
+                FAIL2("panic: reg returned NULL to regatom, flags=%#" UVXf, (UV)flags);
 	}
 	*flagp |= flags&(HASWIDTH|SPSTART|SIMPLE|POSTPONED);
 	break;
@@ -10509,8 +10510,8 @@ tryagain:
                 /* regclass() can only return RESTART_UTF8 if multi-char folds
                    are allowed.  */
                 if (!ret)
-                    FAIL2("panic: regclass returned NULL to regatom, flags=%#X",
-                          *flagp);
+                    FAIL2("panic: regclass returned NULL to regatom, flags=%#" UVXf,
+                          (UV)*flagp);
 
 		RExC_parse--;
 
@@ -11633,8 +11634,8 @@ S_handle_regex_sets(pTHX_ RExC_state_t *pRExC_state, SV** return_invlist, I32 *f
                                   FALSE, /* don't allow multi-char folds */
                                   TRUE, /* silence non-portable warnings. */
                                   &current))
-                        FAIL2("panic: regclass returned NULL to handle_sets, flags=%#X",
-                              *flagp);
+                        FAIL2("panic: regclass returned NULL to handle_sets, flags=%#" UVXf,
+                              (UV)*flagp);
 
                     /* function call leaves parse pointing to the ']', except
                      * if we faked it */
@@ -11799,8 +11800,8 @@ S_handle_regex_sets(pTHX_ RExC_state_t *pRExC_state, SV** return_invlist, I32 *f
                               FALSE, /* don't allow multi-char folds */
                               FALSE, /* don't silence non-portable warnings.  */
                               &current))
-                    FAIL2("panic: regclass returned NULL to handle_sets, flags=%#X",
-                          *flagp);
+                    FAIL2("panic: regclass returned NULL to handle_sets, flags=%#" UVXf,
+                          (UV)*flagp);
                 /* regclass() will return with parsing just the \ sequence,
                  * leaving the parse pointer at the next thing to parse */
                 RExC_parse--;
@@ -11822,8 +11823,8 @@ S_handle_regex_sets(pTHX_ RExC_state_t *pRExC_state, SV** return_invlist, I32 *f
                              FALSE, /* don't allow multi-char folds */
                              FALSE, /* don't silence non-portable warnings.  */
                              &current))
-                    FAIL2("panic: regclass returned NULL to handle_sets, flags=%#X",
-                          *flagp);
+                    FAIL2("panic: regclass returned NULL to handle_sets, flags=%#" UVXf,
+                          (UV)*flagp);
                 /* function call leaves parse pointing to the ']', except if we
                  * faked it */
                 if (is_posix_class) {
-- 
1.7.5.4


```
</details>

**Comment by p5pRT** [comment]
### From @demerphq

Damn\, more from me\. Im definitely going to have buy you beer or
something if you keep this up\. :\-\(

On 2 May 2013 12&#8203;:29\, Robin Barker \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  Robin Barker
> \# Please include the string&#8203;:  \[perl \#117837\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=117837 >
>
>
> This is a bug report for perl from rmbarker\.cpan@&#8203;btinternet\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.17\.12\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> When compiling perl\, I get many \-Wformat warnings from regcomp\.c
> %X format is used but the variable is not an int \(i\.e\. is longer\)
> The normal fix for this is to use UVXf and cast the variable to \(UV\)
> and this is what the attached patch does\.
>
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>      category=core
>      severity=low
> \-\-\-
> Site configuration information for perl 5\.17\.12&#8203;:
>
> Configured by robin at Wed May  1 12&#8203;:59&#8203;:24 BST 2013\.
>
> Summary of my perl5 \(revision 5 version 17 subversion 12\) configuration&#8203;:
>    Commit id&#8203;: 8257347d5f11b9bd1df15345570993538c8b1696
>    Platform&#8203;:
>      osname=linux\, osvers=3\.0\.0\-32\-generic\, archname=i686\-linux\-64int
>      uname='linux king\-ubuntu 3\.0\.0\-32\-generic \#51\-ubuntu smp thu mar 21
> 15&#8203;:51&#8203;:26 utc 2013 i686 athlon i386 gnulinux '
>      config\_args='\-ders \-Dcc=gcc \-Dusedevel \-Doptimize=\-O2 \-DDEBUGGING
> \-Duse64bitint \-Dman3dir=none \-Dcf\_email=rmbarker\.cpan@&#8203;btinternet\.com'
>      hint=recommended\, useposix=true\, d\_sigaction=define
>      useithreads=undef\, usemultiplicity=undef
>      useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
>      use64bitint=define\, use64bitall=undef\, uselongdouble=undef
>      usemymalloc=n\, bincompat5005=undef
>    Compiler&#8203;:
>      cc='gcc'\, ccflags ='\-DDEBUGGING \-fno\-strict\-aliasing \-pipe
> \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE
> \-D\_FILE\_OFFSET\_BITS=64'\,
>      optimize='\-O2 \-g'\,
>      cppflags='\-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector
> \-I/usr/local/include'
>      ccversion=''\, gccversion='4\.6\.4'\, gccosandvers=''
>      intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=12345678
>      d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>      ivtype='long long'\, ivsize=8\, nvtype='double'\, nvsize=8\,
> Off\_t='off\_t'\, lseeksize=8
>      alignbytes=4\, prototype=define
>    Linker and Libraries&#8203;:
>      ld='gcc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
>      libpth=/usr/local/lib /lib/i386\-linux\-gnu /lib/\.\./lib
> /usr/lib/i386\-linux\-gnu /usr/lib/\.\./lib /lib /usr/lib
>      libs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
>      perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
>      libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
>      gnulibc\_version='2\.13'
>    Dynamic Linking&#8203;:
>      dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
>      cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib
> \-fstack\-protector'
>
> Locally applied patches&#8203;:
>
>
> \-\-\-
> @&#8203;INC for perl 5\.17\.12&#8203;:
>      lib
>      /usr/local/lib/perl5/site\_perl/5\.17\.12/i686\-linux\-64int
>      /usr/local/lib/perl5/site\_perl/5\.17\.12
>      /usr/local/lib/perl5/5\.17\.12/i686\-linux\-64int
>      /usr/local/lib/perl5/5\.17\.12
>      \.
>
> \-\-\-
> Environment for perl 5\.17\.12&#8203;:
>      HOME=/home/robin
>      LANG=en\_GB\.UTF\-8
>      LANGUAGE=en\_GB&#8203;:en
>      LD\_LIBRARY\_PATH \(unset\)
>      LOGDIR \(unset\)
> PATH=/usr/lib/lightdm/lightdm&#8203;:/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/sbin&#8203;:/usr/bin&#8203;:/sbin&#8203;:/bin&#8203;:/usr/games
>      PERL\_BADLANG \(unset\)
>      SHELL=/bin/bash



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From rmbarker@cpan.org

On Thu May 02 03&#8203;:48&#8203;:16 2013\, demerphq wrote&#8203;:
> Damn\, more from me\. Im definitely going to have buy you beer or
> something if you keep this up\. :\-\(
> 
Apparently fixed by commit d6dda2f0319879e9dc262810a1d42
Suggest closed

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @khwilliamson

On Sun Jun 09 10&#8203;:46&#8203;:17 2013\, RMB wrote&#8203;:
> On Thu May 02 03&#8203;:48&#8203;:16 2013\, demerphq wrote&#8203;:
> > Damn\, more from me\. Im definitely going to have buy you beer or
> > something if you keep this up\. :\-\(
> > 
> Apparently fixed by commit d6dda2f0319879e9dc262810a1d42
> Suggest closed

Sorry that I didn't see that Robin had submitted a patch\, so I redid it
my self\.  For the record\, the commit number is 
cc5524b47a7d6dda2f0319879e9dc262810a1d42
\-\- 
Karl Williamson
